judge/runguard.c

```diff
@@ -852,6 +860,71 @@ void setrestrictions()
 	}
 }
 
+void pump_pipes(fd_set readfds, size_t data_read[], size_t data_passed[])
+{
+	ssize_t nread;
+	size_t to_read;
+	int i;
+
+	/* Check to see if data is available and pass it on */
+	for(i=1; i<=2; i++) {
+		if ( child_pipefd[i][PIPE_OUT] != -1 &&
+		     FD_ISSET(child_pipefd[i][PIPE_OUT],&readfds) ) {
+
+			if (limit_streamsize && data_passed[i] == streamsize) {
+				/* Throw away data if we're at the output limit, but
+				   still count how much data we consumed  */
+				nread = read(child_pipefd[i][PIPE_OUT], buf, BUF_SIZE);
+			} else {
+				/* Otherwise copy the output to a file */
+				to_read = BUF_SIZE;
+				if (limit_streamsize) {
+					to_read = min(BUF_SIZE, streamsize-data_passed[i]);
+				}
+
+				if ( use_splice ) {
+					nread = splice(child_pipefd[i][PIPE_OUT], NULL,
+					               child_redirfd[i], NULL,
+					               to_read, SPLICE_F_MOVE);
+
+					if ( nread==-1 && errno==EINVAL ) {
+						use_splice = 0;
+						verbose("splice failed, switching to read/write");
+						/* Setting errno here to repeat the copy. */
+						errno = EAGAIN;
+					}
+				} else {
+					nread = read(child_pipefd[i][PIPE_OUT], buf, to_read);
+					if ( nread>0 ) {
+						nread = write(child_redirfd[i], buf, nread);
+					}
+				}
+
+				if ( nread>0 ) data_passed[i] += nread;
+
+				/* print message if we're at the streamsize limit */
+				if (limit_streamsize && data_passed[i] == streamsize) {
+					verbose("child fd %i limit reached",i);
+				}
+			}
+			if ( nread==-1 ) {
+				if (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK) continue;
+				error(errno,"copying data fd %d",i);
+			}
+			if ( nread==0 ) {
+				/* EOF detected: close fd and indicate this with -1 */
+				if ( close(child_pipefd[i][PIPE_OUT])!=0 ) {
+					error(errno,"closing pipe for fd %d",i);
+				}
+				child_pipefd[i][PIPE_OUT] = -1;
```

Comment:
This is where we detect EOF. After pumping all data, these should be -1, right?
(See next comment.)